tsdb/chunks/chunks.go

```diff
@@ -445,28 +480,50 @@ func (s *Reader) Size() int64 {
 // Chunk returns a chunk from a given reference.
 func (s *Reader) Chunk(ref uint64) (chunkenc.Chunk, error) {
 	var (
-		sgmSeq    = int(ref >> 32)
-		sgmOffset = int((ref << 32) >> 32)
+		// Get the upper 4 bytes.
+		// These contain the segment index.
+		sgmIndex = int(ref >> 32)
+		// Get the lower 4 bytes.
+		// These contain the segment offset where the data for this chunk starts.
+		sgmChunkOffset = int((ref << 32) >> 32)
 	)
-	if sgmSeq >= len(s.bs) {
-		return nil, errors.Errorf("reference sequence %d out of range", sgmSeq)
+
+	if sgmIndex >= len(s.bs) {
+		return nil, errors.Errorf("segment index %d out of range", sgmIndex)
 	}
-	chkS := s.bs[sgmSeq]
 
-	if sgmOffset >= chkS.Len() {
-		return nil, errors.Errorf("offset %d beyond data size %d", sgmOffset, chkS.Len())
+	chkBytes := s.bs[sgmIndex]
+
+	if sgmChunkOffset+maxChunkLengthFieldSize > chkBytes.Len()-SegmentHeaderSize {
+		return nil, errors.Errorf("segment doesn't include enough bytes to read the chunk size data field - required:%v, available:%v", sgmChunkOffset+maxChunkLengthFieldSize+SegmentHeaderSize, chkBytes.Len())
```

Comment:
I think this should be 
```
if sgmChunkOffset+maxChunkLengthFieldSize > chkBytes.Len() {
```

Reply:
Aaah actually this needs to be 
```
if sgmChunkOffset+maxChunkLengthFieldSize > chkBytes.Len()+SegmentHeaderSize {
```

`chkBytes.Len()` doesn't include the size of the segment header. The minimum offset is always 8bytes which is the size of the segment header and need to account for that.
Here is the calculation with an offset 8.
8+5 > 5+8

It will hit the error if the chunk bytes are 4 which is not enough to read the chunks length field

Reply:
aaah I just double checked and you are right `chkBytes.Len()` does include the segment header so your calculation is correct. Thanks